xip/xipether.c

```diff
@@ -0,0 +1,449 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <sys/types.h>
+#include <asm/byteorder.h>
+#include <asm-generic/errno-base.h>
+#include <net/xia_fib.h>
+#include <xia_socket.h>
+#include <sys/socket.h>
+#include <linux/netdevice.h>
+
+#include "xip_common.h"
+#include "utils.h"
+#include "ll_map.h"
+#include "libnetlink.h"
+#include "xiart.h"
+
+#define IFINDEX_STR_SIZE 8
+#define XIA_LLADDR_LEN 12
+
+static int usage(void)
+{
+	fprintf(stderr,
+"Usage: xip ether { addif | delif } dev DEV\n"
+"       xip ether { addneigh | delneigh } lladdr LLADDR dev DEV\n"
+"       xip ether show { interfaces | neighs }\n"
+"where  LLADDR := HEXDIGIT{1,2} (':' HEXDIGIT{1,2})*\n"
+"       DEV := STRING NUMBER\n");
+	return -1;
+}
+
+static int form_ether_xid(unsigned int oif, unsigned char *lladdr,
+			  unsigned int tlen, char *id)
+{
+	char bytes[4];
+	int rc;
+
+	bytes[0] = (oif >> 24) & 0xFF;
+	bytes[1] = (oif >> 16) & 0xFF;
+	bytes[2] = (oif >> 8) & 0xFF;
+	bytes[3] = oif & 0xFF;
+	rc = snprintf(id, tlen,
+			"%02x%02x%02x%02x%02x%02x%02x%02x%02x%02x%020x",
+			(unsigned char)bytes[0], (unsigned char)bytes[1],
+			(unsigned char)bytes[2], (unsigned char)bytes[3],
```

Comment:
Variable `bytes` is just being used to put the value of variable `oif` into big endian. Instead, (1) drop variable `bytes`, (2) replace the beginning part of the format, namely "%02x%02x%02x%02x", with "%08x", and (3) pass `__cpu_to_be32(oif)` as the parameter to this format.